appserver/web/war-util/src/test/java/org/glassfish/web/loader/WebappClassLoaderTest.java

```diff
@@ -0,0 +1,178 @@
+ * its licensees as provided above.  However, if you add GPL Version 2 code
+ * and therefore, elected the GPL Version 2 license, then the option applies
+ * only if the new code is made subject to such option by the copyright
+ * holder.
+ */
+package org.glassfish.web.loader;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CyclicBarrier;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+
+import org.apache.naming.resources.FileDirContext;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+public class WebappClassLoaderTest {
+
+    private static CyclicBarrier lock;
+    private static ExecutorService executor;
+    private static File junitJarFile;
+
+    @BeforeClass
+    public static void setup() throws URISyntaxException {
+        // Run 3 methods at the same time, and make the pool large enough to increase
+        // the chance of a race condition
+        lock = new CyclicBarrier(3);
+        executor = Executors.newFixedThreadPool(60);
+
+        // Fetch any JAR to use for classloading
+        junitJarFile = new File(Test.class.getProtectionDomain().getCodeSource().getLocation().toURI());
+    }
+
+    @AfterClass
+    public static void shutdown() {
+        if (executor != null) {
+            executor.shutdownNow();
+        }
+    }
+
+    @Test
+    public void check_findResourceInternalFromJars_thread_safety() throws Exception {
+        final ClassLoader classLoader = this.getClass().getClassLoader();
+        final WebappClassLoader webappClassLoader = new WebappClassLoader(classLoader, null);
+        webappClassLoader.start();
+        webappClassLoader.setResources(new FileDirContext());
+
+        CompletableFuture<Exception> result = new CompletableFuture<>();
+
+        // Create the tasks, and have them each run at the same time
+        // using the cyclic barrier
+        Runnable lookupTask = waitAndDo(lock, result, () -> lookup(classLoader, webappClassLoader));
+        Runnable addTask = waitAndDo(lock, result, () -> add(classLoader, webappClassLoader));
+        Runnable closeTask = waitAndDo(lock, result, () -> webappClassLoader.closeJARs(true));
+
+        try {
+            // Run the methods at the same time
+            for (int i = 0; i < 100; i++) {
+                executor.execute(addTask);
+                executor.execute(lookupTask);
+                executor.execute(closeTask);
+            }
+            // Check to see if any completed exceptionally
+            Exception ex = result.get(200, TimeUnit.MILLISECONDS);
```

Comment:
This also mean that all tasks must be finished 200 ms after the lasts task is scheduled.

I tested this on my machine and the time between the end of the last task finished and the timeout was only about 35 ms.

So this should help 

            boolean stopped = executor.awaitTermination(500, TimeUnit.MILLISECONDS);
            if (!stopped) {
                fail("Tasks are not finished in a timely manner");
            }

But it revealed another issue, all the tasks are not finished when retrieving the Future.get() :  the lookup task never ends and seems blocked in a deadlock.

Reply:
> Running the test on my system suggests that there is a deadlock.

I had to change values, but I can cause the deadlock too under some circumstances ...

Reply:
> This also mean that all tasks must be finished 200 ms after the lasts task is scheduled.

It does not, it waits only to first `result.complete(ex)` call.

Reply:
It seems not to be a deadlock, but the lookup tasks just takes to long by the use of

    Thread.sleep(0, 100);

So that lookup tasks are not finished.

By replacing it with

    Thread.yield();

Tests runs normal and all tasks are finished 500 ms after the last one is launched (with .awaitTermination() )

Reply:
With more threads it still does not finish ... I will try yet to use jcmd if it would see the deadlock or not, but even my cpu load dropped after a while far before timeout.